core/drivers/crypto/asu_driver/asu_hash.c

```diff
@@ -0,0 +1,444 @@
+
+	if (!len) {
+		DMSG("This is 0 len task, skip");
+		return TEE_SUCCESS;
+	}
+
+	if (!data && len) {
+		EMSG("Invalid input parameters");
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	asu_hashctx = to_hash_ctx(ctx);
+	if (asu_hashctx->uniqueid == ASU_UNIQUE_ID_MAX)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	return asu_hash_update(asu_hashctx, (uint8_t *)data, len);
+}
+
+static TEE_Result asu_hash_cb(void *cbrefptr, struct asu_resp_buf *resp_buf)
+{
+	struct asu_hash_cbctx *cbctx = NULL;
+	uint8_t *src_addr = NULL;
+
+	cbctx = cbrefptr;
+	src_addr = (uint8_t *)&resp_buf->arg[ASU_RESPONSE_BUFF_ADDR_INDEX];
+	memcpy(cbctx->digest, src_addr, cbctx->len);
+
+	return TEE_SUCCESS;
+}
+
+/**
+ * asu_hash_final() - Send final request to engine.
+ * @asu_hashctx:Request private hash context
+ * @digest:	Output digest buffer
+ * @len:	Size of digest buffer
+ *
+ * Send final request to engine and populate digest result
+ *
+ * Return: TEE_SUCCESS, TEE_ERROR_BAD_PARAMETERS or TEE_ERROR_GENERIC
+ */
+
+static TEE_Result asu_hash_final(struct asu_hash_ctx *asu_hashctx,
+				 uint8_t *digest, size_t len)
+{
+	TEE_Result ret = TEE_SUCCESS;
+	struct asu_sha_op_cmd op = {};
+	struct asu_client_params *cparam = NULL;
+	struct asu_hash_cbctx cbctx = {};
+
+	if (!digest || len == 0)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	cbctx.digest = digest;
+	cbctx.len = len;
+	cparam = &asu_hashctx->cparam;
+	cparam->priority = ASU_PRIORITY_HIGH;
+	cparam->cbptr = &cbctx;
+	cparam->cbhandler = asu_hash_cb;
+
+	/* Inputs of SHA request */
+	op.dataaddr = 0;
+	op.datasize = 0;
+	op.hashaddr = virt_to_phys((void *)digest);
+	op.hashbufsize = len;
+	if (asu_hashctx->shamode == ASU_SHA_MODE_SHA256)
+		op.hashbufsize = ASU_SHA_256_HASH_LEN;
+	else if (asu_hashctx->shamode == ASU_SHA_MODE_SHA384)
+		op.hashbufsize = ASU_SHA_384_HASH_LEN;
+	else if (asu_hashctx->shamode == ASU_SHA_MODE_SHA512)
+		op.hashbufsize = ASU_SHA_512_HASH_LEN;
+
+	op.shamode = asu_hashctx->shamode;
+	op.islast = 1;
+	op.opflags = ASU_SHA_FINISH | asu_hashctx->shastart;
+	ret = asu_sha_op(asu_hashctx, &op, asu_hashctx->module);
+	cache_operation(TEE_CACHEFLUSH, digest, op.hashbufsize);
+
+	return ret;
+}
+
+static TEE_Result asu_hash_do_final(struct crypto_hash_ctx *ctx,
+				    uint8_t *digest, size_t len)
+{
+	struct asu_hash_ctx *asu_hashctx = NULL;
+
+	asu_hashctx = to_hash_ctx(ctx);
+
+	return asu_hash_final(asu_hashctx, digest, len);
+}
+
+/**
+ * asu_hash_ctx_free() - Free Private context.
+ * @crypto_hash_ctx: crypto context used by the crypto_hash_*() functions
+ * Release crypto engine and free private context memory.
+ *
+ * Return: void
+ */
+
+static void asu_hash_ctx_free(struct crypto_hash_ctx *ctx)
+{
+	struct asu_hash_ctx *asu_hashctx = NULL;
+
+	asu_hashctx = to_hash_ctx(ctx);
+	asu_free_unique_id(asu_hashctx->uniqueid);
+	asu_hashctx->uniqueid = ASU_UNIQUE_ID_MAX;
+	mutex_lock(&asu_shadev->engine_lock);
+	if (asu_hashctx->module == ASU_MODULE_SHA2_ID &&
+	    !asu_shadev->sha2_available)
```

Comment:
If `asu_hashctx->module == ASU_MODULE_SHA2_ID` then `asu_shadev->sha2_available` must be false or there's an inconsistency. There's no need to check `asu_shadev->sha2_available`. If we must, then an `assert()` before setting it to true makes more sense.